src/Columns/IColumnImpl.h

```diff
@@ -139,4 +139,41 @@ bool IColumn::hasEqualValuesImpl() const
     return true;
 }
 
+
+template <typename Derived>
+double IColumn::getRatioOfDefaultRowsImpl(double sample_ratio) const
+{
+    assert(sample_ratio > 0 && sample_ratio <= 1.0);
+
+    size_t num_rows = size();
+    size_t num_sampled_rows = static_cast<size_t>(num_rows * sample_ratio);
+    if (num_sampled_rows == 0)
+        return 0.0;
+
+    size_t step = num_rows / num_sampled_rows;
+    std::uniform_int_distribution<size_t> dist(1, step);
```

Comment:
1. For sample_ratio = 1 we don't need any distribution - maybe write separate implementation
2. Do we really need random? It could be slow. I would just take each step'th or something
3. Actually I don't like step too. For `ratio > 1/2` `step < 2/1` and actually is 1. For example, we can take `i`'ts row if `num_checked_rows / i < sample_ratio <=> num_checked_rows * num_rows < i * num_sampled_rows`

Reply:
Fixed. I've added a search pattern, that you suggest, but with a randomization of row of search start.